resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -1053,16 +1066,29 @@ private void tryToFinishResolve(final DnsServerAddressStream nameServerAddrStrea
 
             // .. and we could not find any expected records.
 
-            // If cause != null we know this was caused by a timeout / cancel / transport exception. In this case we
-            // won't try to resolve the CNAME as we only should do this if we could not get the expected records
-            // because they do not exist and the DNS server did probably signal it.
-            if (cause == null && !triedCNAME &&
-                    (question.type() == DnsRecordType.A || question.type() == DnsRecordType.AAAA)) {
-                // As the last resort, try to query CNAME, just in case the name server has it.
-                triedCNAME = true;
-
-                query(hostname, DnsRecordType.CNAME, getNameServers(hostname), true, promise);
-                return;
+            // The following is of questionable benefit, but has been around for a while that
+            // it may be risky to remove.
+            // - If we receive NXDOMAIN we know the domain doesn't exist, any other lookup type is meaningless.
+            // - If we receive SERVFAIL, and we attempt a CNAME that returns NOERROR with 0 answers, it may lead the
+            //   call-site to invalidate previously advertised addresses.
+            // Having said that, there is the case of DNS services that don't respect the protocol either
+            // - An A lookup could result in NXDOMAIN but a CNAME may succeed with answers.
+            // It's an imperfect world. Accept it.
```

Comment:
@tkountis can you point me to the relevant RFC section ?

Reply:
Sure [NXDOMAIN: There Really Is Nothing Underneath](https://datatracker.ietf.org/doc/html/rfc8020)

Reply:
@tkountis thanks... please add to the comment. Also I noticed it's only "proposed" for now.

Reply:
Updated the comment to include RFC.
Yes its "unfortunately" proposed, but most servers are already respecting it. 
There could still be some legacy old software that doesn't, but i dont think this piece of code offers much value on that front either.

In my minds eye, this looks like an optimization, possibly related to https://www.rfc-editor.org/rfc/rfc1034#section-3.6.2 
> CNAME RRs cause special action in DNS software.  When a name server
fails to find a desired RR in the resource set associated with the
domain name, it checks to see if the resource set consists of a CNAME
record with a matching class.  If so, the name server includes the CNAME
record in the response and restarts the query at the domain name
specified in the data field of the CNAME record.  The one exception to
this rule is that queries which match the CNAME type are not restarted.

but even in this case, this is a specialization for the nameserver side, not a resolver.